README
 * This is modified work by Miroslav Madzarevic githubcode@osadmin.com as of May 1st 2002
 * based originally on proxy-2.2.4 program by sparlin@openpro.org

Be sure to read file INSTALL

Proxy load balance is a C program written to balance incoming connections
to numerous end boxes (computers) using a simple weight algorithm. With its low memory
and disk footprint, threads, it is ideal for professional load balancing.

It uses a single config file (by default config.txt in the current directory) 
where all the setup is made. That file contains on each line box name or ip address, 
box tcp port and weight of that box. Weight is something you write yourself according  to the power 
of the end box. The stronger the box is (more ram, faster cpu, faster internet connection)
the more weight it has and it is more probable that the next outbound connection will use it.

You can change config.txt dinamically with the program running. proxylb reads config.txt on
each connection so it will see the changes. I plan on writing a bundle of perl scripts for 
box monitoring that will update this file, maybe even create pipe or use database to get
current connections.

The program is written in C, compiled on my box has some 13 kilobytes and uses threads.
It is webmasters choice of software although it has many more uses.

It can be used to balance any kind of tcp traffic including and not limited to http, smtp,
pop3, ssl, ftp (passive, maybe even active). It can be used for simple port binding, for example
to make your internal (on LAN) httpd available to the world by linking it onto your firewall 
box.

READ CHANGELOG file to see what has changed (it has more info than what was presented here).

READ THE INSTALL INSTRUCTIONS IN THE FILE LABELED INSTALL.

You can read proxy original readme below (Hint: it is probably not necessary for you to read it).  

INTRODUCTION 
Proxy is a C application that when run on a multi-homed host will forward
all packets from source, to destination. Where source might be a system on
the Internet, and destination might be a box on a private network behind a
multi-homed Linux machine. The way I use it is to change the telnet port on
a machine on my internal network from 23 to 999, then I'll run the proxy on
the multi-homed box, and from the Internet, telnet to the multi-homed box
on port 999 and it will automagically put me on the internal machine.

client			proxy_server		Telnet Server
+----------+		+----------+		+----------+
| Internet |- - - >	|     |    |- - - >	| Real     |
| Host	   | telnet	|NIC1 |NIC2|     23	| telnet   |
|          |  999	|     |    |		| server   |
+----------+		+----------+		+----------+
0.0.0.0/0      		192.*   10.*    	10.*

The arguments to proxy when run on the multi-homed host would be:

proxy -s 999 -D internal.private.net -d 23

For a complete list of options to proxy, run 'proxy --help'.

This tells proxy to accept connections on port 999 and send them to the
internal machine on port 23. This example shows the power of proxy, you can
also use it to set up an internal web server, or mail server, etc, etc...
without having any of the machines connected to the net except for the
multi-homed box.

You can use an optional 4-th parameter (-S) to tell proxy which one of
machine's assigned IP's bind to. If this is omitted, proxy will bind to all
of the addresses. This may be useful if you want the proxied port to only
be visible to localhost or to serve requests for different IP addresses
(but the same port) differently. This will also let you choose a different
destination port.

FILTERING 
Proxy-2.2.4 can also do filtering. In order to take advantage of the
filtering rules of proxy-2.2, you must make a config file. The file will
reside in the /etc directory and it will be called "proxy.filters"
(i.e. /etc/proxy.filters).  The filter file can contain comments, only if
the first character of the line starts with the pound sign '#'. Example:

Example 1:
# This is a proper comment. (correct)

Example 2:
DENY.209.209.123.12 # DENY this host (Not correct)

The first example is a correct comment, the second isn't. 

The filters file can be setup with a default policy for either accept or
deny.  If a default policy of deny is chosen, then all traffic that hasn't
been specifically allowed, will be denied. By contrast, if a default policy
of accept is given, all traffic that isn't specifically denied will be
allowed.

proxy filters are to be written as 'rule.ip', here is an example:

ACCEPT.127.0.0.1
DENY.127.0.0.1

Please note that names are not allowed, so:

DENY.localhost 

will _not_ work.

There is no order in which filter rules have to be written, this means that
the default policy could be at the top, middle or bottom of the file. The
default policy must be written on a line by itself terminated by a
semi-colon, example:

ACCEPT;

or

DENY; 

Alternatively, if there is no filters file in place, all connections will
be allowed. Please see the file called "proxy.filters_example" in this
directory for more detail.

CHANGES 
Proxy 2.2.4 changes include: 
	+ Use pthread_detach to detach threads...
	+ For some reason, creating your thread in a detached state is not
	the same as calling pthread_detach() after calling pthread_create().
	I can now create more than 1024 threads, things are good again...
	+ I have written a small Perl app that can be used to stress test
	proxy. Basically, you call it like so:

	./proxy -s <port> -d 80 -D some.web.server -n
	./regress server port #

	where server is the server running proxy, port is the port proxy is
	listening on, and # is the number of connections you want to make to
	proxy.

	Things are very stable now, and proxy can be considered for use in
	production environments.

AUTHOR
Proxy was written and is maintained by Sonny Parlin.

CONTRIBUTERS
Dmitri E. Kalintsev: dek@hades.uz
Added extra parameter for specifying source IP for binding.

LICENSE 
This program is free software; you can redistribute it and/or modify it
under the terms of the GNU General Public License as published by the Free
Software Foundation; either version 2 of the License, or any later version.

This program is distributed in the hope that it will be useful, but WITHOUT
ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
more details.

You should have received a copy of the GNU General Public License along
with this program; if not, write to the Free Software Foundation, Inc., 59
Temple Place - Suite 330, Boston, MA 02111-1307, USA.


CHANGELOG
version 1.0.3
    proxylb runs in foreground and in background without any problems
    all logging is done via syslog only when -d (debug) flag is set on the command line
    almost nothing (except for the one time startup messages) is printed on stdout
    code looks much better (I reformated it) :)

version 1.0.2
    Fixed parse problems on 2.9x gcc (now compiles on suse)
    proxylb runs only in foreground (quick fix for background deamon craching)



list.c
#include "proxylb.h"

// inserts node at a start of a list
int insert_node( char *host_name, int port, int *ptr_nr_nodes, struct node **ptr_head, int debug ) {
    struct node *ptr;

    // debug==1?syslog(LOG_INFO,"insert_node got host_name '%s' port '%d'",host_name,port):NULL;

    /* we don't need to debug so hard, do we ? :)

    if (debug == 1) {
        if ((*ptr_head)==NULL) {
           syslog(LOG_INFO,"head is NULL, number of elements is %d",(*ptr_nr_nodes));
        } else {
           syslog(LOG_INFO,"head is NOT null, number of elements is %d",(*ptr_nr_nodes));
        }
    }

    */
 
    ptr = (struct node *) calloc( 1, sizeof(struct node ) );
    if( ptr == NULL ) {                     /* error allocating node?      */
        exit(1);                            /*   exit, else                */
    } else {                                /* allocated node successfully */
       // make host_name string
       ptr->host_name = (char *) calloc(strlen(host_name)+1,sizeof(char));
       strcpy(ptr->host_name,host_name);
       ptr->port   = port;                  /* copy port, set head         */
       ptr->next   = (struct node *) (*ptr_head);
       (*ptr_head) = ptr;
       ++(*ptr_nr_nodes);                   /* ++ number, return true      */ 
   }

   // debug==1?syslog(LOG_INFO,"insert_node end of function"):NULL;
 
   return 1;                         
 
}

// gets node by integer position
// position 1 is the first position
struct node *get_node( int position, int nr_nodes, struct node *head ) {

    struct node *ptr=head;    

    if ( (position>nr_nodes) || (position<1) ) return NULL;
    
    while ( (ptr != NULL) && (--position) ) {
       
        ptr = ptr->next;
        
    }

    return ptr;
}


// destroys list
int kill_list(struct node **ptr_head, int *ptr_nr_nodes, int debug) {

    struct node *ptr=(*ptr_head);
    struct node *tmp;    
    int counter=0;

    
    while ( ptr != NULL ) {
        debug==1?syslog(LOG_INFO,"--- freeing list, pass %d",++counter):NULL;
        tmp = ptr;
        ptr = ptr->next;
        free((char *) tmp->host_name);
        free((struct node *) tmp);
        
    }

    (*ptr_head)     = NULL;
    (*ptr_nr_nodes) = 0;

    return 1;
}




util.c
#include "proxylb.h"

// gets random conection
// allocates ip and port
int get_random_connection_params(char **_ip, char **_port, struct node *head, int nr_nodes, int debug ) {
    int rn;
    struct node *ptr;
    int nr_printed;

    // generate random number

    rn = 1+(int) ((double)(nr_nodes)*rand()/(RAND_MAX+1.0));

    // get random connection

    ptr = get_node(rn,nr_nodes,head);

    // modify connection parameters
    
    (*_ip)   = (char *) calloc(strlen(ptr->host_name)+1,sizeof(char));
    strcpy((*_ip),ptr->host_name);    
    (*_port) = (char *) calloc(6,sizeof(char));
    nr_printed = snprintf((*_port),6,"%d",ptr->port);
    (*_port)[5]='\0';
    // debug==1?syslog(LOG_INFO,"Number of characters: %d, port is '%s'",nr_printed,(*_port)):NULL;

    
    // return success
    return 1;
}




config.c
#include "proxylb.h"


int read_config(struct node **ptr_head, int *ptr_nr_nodes, char *conf_file, int debug) {

    FILE   *fp;          // declare file pointer 
    int    port, weight;
    char   *host_name; 
    // struct node *ptr;
    int    i=0;
    
    (*ptr_nr_nodes) = 0;
    (*ptr_head)     = NULL;

    fp = fopen ((conf_file==NULL)?"./config.txt":conf_file, "r");  /* open the file for reading */

    if (fp == NULL) {
        debug==1?syslog(LOG_ERR, "can't open input file '%s'!",
            (conf_file==NULL)?"./config.txt":conf_file):NULL;
        exit(1);
    }

    // string host_name is automatically allocated
    while (fscanf(fp, "%as %d %d", &host_name, &port, &weight) != EOF) {
       
        debug==1?syslog(LOG_INFO, "'%s' %d %d", host_name, port, weight):NULL;

        
        for(i=0;i<weight;i++) {
            debug==1?syslog(LOG_INFO, "--- inserting new node for host_name '%s', weight:%d", host_name, weight):NULL;
            insert_node(host_name,port,ptr_nr_nodes,ptr_head,debug);
        }

        
        free((char *) host_name); // free the host_name string

        
    }

    fclose(fp);  


    // debugging
    /* we don't have to debug that hard ;)
    if (debug==1) {
        for(i=0;i<(*ptr_nr_nodes);i++) {
            ptr = get_node(i+1,(*ptr_nr_nodes),(*ptr_head));    
            syslog(LOG_INFO, "number %4d, host_name: '%s', port: '%d'", i+1, ptr->host_name, ptr->port);
        }
    }
    */

    return 1;    
}



proxylb.h

/*

 * This is modified work by Miroslav Madzarevic, githubcode@osadmin.com as of May 1st 2002
 * based originally on proxy-2.2.4 program by sparlin@openpro.org

 * COPYRIGHT (C) 2002 MIROSLAV MADZAREVIC, githubcode@osadmin.com
 * COPYRIGHT (C) 1998, 1999, 2000  SONNY PARLIN
 * 
 * THIS PROGRAM IS FREE SOFTWARE; YOU CAN REDISTRIBUTE IT AND/OR
 * MODIFY IT UNDER THE TERMS OF THE GNU GENERAL PUBLIC LICENSE
 * AS PUBLISHED BY THE FREE SOFTWARE FOUNDATION; EITHER VERSION 2
 * OF THE LICENSE, OR ANY LATER VERSION.
 *
 * THIS PROGRAM IS DISTRIBUTED IN THE HOPE THAT IT WILL BE USEFUL,
 * BUT WITHOUT ANY WARRANTY; WITHOUT EVEN THE IMPLIED WARRANTY OF
 * MERCHANTABILITY OR FITNESS FOR A PARTICULAR PURPOSE.  SEE THE
 * GNU GENERAL PUBLIC LICENSE FOR MORE DETAILS.
 *
 * YOU SHOULD HAVE RECEIVED A COPY OF THE GNU GENERAL PUBLIC LICENSE
 * ALONG WITH THIS PROGRAM; IF NOT, WRITE TO THE FREE SOFTWARE
 * FOUNDATION, INC., 59 TEMPLE PLACE - SUITE 330, BOSTON, MA  
 * 02111-1307, USA.

 */


/*-----------------------------------------------*/
/* Universal header for network programming code */
/* Sonny Parlin 12/98				             */
/*-----------------------------------------------*/

#include <stdio.h>
#include <unistd.h>
#include <stdlib.h>
#include <sys/types.h>
#include <sys/socket.h>
#include <netinet/in.h>
#include <pthread.h>
#include <errno.h>
#include <signal.h>
#include <syslog.h>
#include <netdb.h>
#include <fcntl.h>
#include <sys/time.h>
#include <string.h>
#include <arpa/inet.h>
#include <sys/stat.h>
#include <sys/wait.h>
#include <signal.h>
#include <ctype.h>
#include <getopt.h>

#define PROXYLB_VERSION "proxylb-1.0.3"

/* Define bzero() as a macro incase it's not in std C library */
#define bzero(ptr,n) memset(ptr, 0, n)

/* Miscellaneous constants */
#define MAXLINE         4096    /* max text line length */
#define MAXSOCKADDR  	128     /* max socket address structure size */
#define BUFFSIZE        8192    /* buffer size for reads and writes */

/* uncomment next line for non-posix compliant systems */
#define socklen_t unsigned int

/* Following shortens all the type casts of pointer arguments */
#define SA        struct sockaddr

#define LISTENQ 1024 

// mire
/* definition of a data node holding host information */
struct node {
    char   *host_name;
    int    port;
    struct node *next;
};

/* Function declorations for proxy */
void Close(int descriptor);
void PrintUsage(int n);
void *proxy(void *ptr); // main proxy function
int  GetLargestSocket(int sd1, int sd2);
int  daemon_init(const char *pname); // makes our program an unix daemon
int  FilterIP(char *);  // we have our own filtering if you don't use ipchains or similar
void finish(int sig);   // exit pthread
int  tcp_connect(const char *, const char *);
int  tcp_listen(const char *, const char *, socklen_t *);
char *sock_ntop(const SA *, socklen_t);
char *Sock_ntop(const SA *, socklen_t);

// mire

//      list.c
int         insert_node ( char *host_name, int port, int *ptr_nr_nodes, struct node **ptr_head, int debug );
struct node *get_node   ( int position, int nr_nodes, struct node *head );
int         kill_list   ( struct node **ptr_head, int *ptr_nr_nodes, int debug );

//      config.c
int read_config ( struct node **ptr_head, int *ptr_nr_nodes, char *conf_file, int debug );

//      util.c
int get_random_connection_params( char **_ip, char **_port, struct node *head, int nr_nodes, int debug );





sock_ntop.c
/*

 * This is modified work by Miroslav Madzarevic githubcode@osadmin.com as of May 1st 2002
 * based originally on proxy-2.2.4 program by sparlin@openpro.org

 * COPYRIGHT (C) 1998, 1999, 2000  SONNY PARLIN
 *
 * THIS PROGRAM IS FREE SOFTWARE; YOU CAN REDISTRIBUTE IT AND/OR
 * MODIFY IT UNDER THE TERMS OF THE GNU GENERAL PUBLIC LICENSE
 * AS PUBLISHED BY THE FREE SOFTWARE FOUNDATION; EITHER VERSION 2
 * OF THE LICENSE, OR ANY LATER VERSION.
 *
 * THIS PROGRAM IS DISTRIBUTED IN THE HOPE THAT IT WILL BE USEFUL,
 * BUT WITHOUT ANY WARRANTY; WITHOUT EVEN THE IMPLIED WARRANTY OF
 * MERCHANTABILITY OR FITNESS FOR A PARTICULAR PURPOSE.  SEE THE
 * GNU GENERAL PUBLIC LICENSE FOR MORE DETAILS.
 *
 * YOU SHOULD HAVE RECEIVED A COPY OF THE GNU GENERAL PUBLIC LICENSE
 * ALONG WITH THIS PROGRAM; IF NOT, WRITE TO THE FREE SOFTWARE
 * FOUNDATION, INC., 59 TEMPLE PLACE - SUITE 330, BOSTON, MA  
 * 02111-1307, USA.

 */

#include	"proxylb.h"

char *
sock_ntop (const struct sockaddr *sa, socklen_t salen) {
    char        portstr[7];
    static char str[128];		/* Unix domain is largest */

    struct sockaddr_in *sin = (struct sockaddr_in *) sa;

    if (inet_ntop (AF_INET, &sin->sin_addr, str, sizeof (str)) == NULL)
        return (NULL);

    if (ntohs (sin->sin_port) != 0) {
        snprintf (portstr, sizeof (portstr), ".%d", ntohs (sin->sin_port));
        strcat (str, portstr);
    }
    return (str);
}

char *
Sock_ntop (const struct sockaddr *sa, socklen_t salen) {
    char *ptr;

    if ((ptr = sock_ntop (sa, salen)) == NULL) {
        syslog (LOG_ERR,"sock_ntop error %m");	/* inet_ntop() sets errno */
        exit (1);
    }

    return (ptr);
}


tcp_listen.c
/*

 * This is modified work by Miroslav Madzarevic githubcode@osadmin.com as of May 1st 2002
 * based originally on proxy-2.2.4 program by sparlin@openpro.org
   
 * COPYRIGHT (C) 1998, 1999, 2000  SONNY PARLIN
 *
 * THIS PROGRAM IS FREE SOFTWARE; YOU CAN REDISTRIBUTE IT AND/OR
 * MODIFY IT UNDER THE TERMS OF THE GNU GENERAL PUBLIC LICENSE
 * AS PUBLISHED BY THE FREE SOFTWARE FOUNDATION; EITHER VERSION 2
 * OF THE LICENSE, OR ANY LATER VERSION.
 *
 * THIS PROGRAM IS DISTRIBUTED IN THE HOPE THAT IT WILL BE USEFUL,
 * BUT WITHOUT ANY WARRANTY; WITHOUT EVEN THE IMPLIED WARRANTY OF
 * MERCHANTABILITY OR FITNESS FOR A PARTICULAR PURPOSE.  SEE THE
 * GNU GENERAL PUBLIC LICENSE FOR MORE DETAILS.
 *
 * YOU SHOULD HAVE RECEIVED A COPY OF THE GNU GENERAL PUBLIC LICENSE
 * ALONG WITH THIS PROGRAM; IF NOT, WRITE TO THE FREE SOFTWARE
 * FOUNDATION, INC., 59 TEMPLE PLACE - SUITE 330, BOSTON, MA  
 * 02111-1307, USA.

 */

#include "proxylb.h"

int
tcp_listen (const char *host, const char *serv, socklen_t * addrlenp) {
    int       listenfd, n;
    const int on = 1;
    struct    addrinfo hints, *res, *ressave;

    bzero (&hints, sizeof (struct addrinfo));
    hints.ai_flags = AI_PASSIVE;
    hints.ai_socktype = SOCK_STREAM;

    if ((n = getaddrinfo (host, serv, &hints, &res)) != 0) {
        syslog(LOG_ERR, "tcp_listen error for host '%s', serv '%s'", host, serv);
    }

    ressave = res;

    do {
        listenfd = socket (AF_INET, res->ai_socktype, res->ai_protocol);
        if (listenfd < 0)
        	continue;

        if (setsockopt (listenfd, SOL_SOCKET, SO_REUSEADDR, &on, sizeof (on)) != 0) {
            syslog(LOG_ERR, "setsockopt error %m");
        }

        if (bind (listenfd, res->ai_addr, res->ai_addrlen) == 0)
            break;      /* success */

        close (listenfd);
    }  while ((res->ai_next) != NULL);

    if (res == NULL) {
      syslog (LOG_ERR, "tcp_listen error for host '%s', serv '%s'", host, serv);
      exit (1);
    }

    if (listen (listenfd, LISTENQ) != 0) {
        syslog(LOG_ERR, "listen error %m");
    }

    if (addrlenp)
        *addrlenp = res->ai_addrlen;	/* return size of protocol address */

    freeaddrinfo (ressave);

    return listenfd;
}


proxylb.c
/*

 * This is modified work by Miroslav Madzarevic githubcode@osadmin.com as of May 1st 2002
 * based originally on proxy-2.2.4 program by sparlin@openpro.org

 * COPYRIGHT (C) 2002 MIROSLAV MADZAREVIC, githubcode@osadmin.com
 * COPYRIGHT (C) 1998, 1999, 2000  SONNY PARLIN
 *
 * THIS PROGRAM IS FREE SOFTWARE; YOU CAN REDISTRIBUTE IT AND/OR
 * MODIFY IT UNDER THE TERMS OF THE GNU GENERAL PUBLIC LICENSE
 * AS PUBLISHED BY THE FREE SOFTWARE FOUNDATION; EITHER VERSION 2
 * OF THE LICENSE, OR ANY LATER VERSION.
 *
 * THIS PROGRAM IS DISTRIBUTED IN THE HOPE THAT IT WILL BE USEFUL,
 * BUT WITHOUT ANY WARRANTY; WITHOUT EVEN THE IMPLIED WARRANTY OF
 * MERCHANTABILITY OR FITNESS FOR A PARTICULAR PURPOSE.  SEE THE
 * GNU GENERAL PUBLIC LICENSE FOR MORE DETAILS.
 *
 * YOU SHOULD HAVE RECEIVED A COPY OF THE GNU GENERAL PUBLIC LICENSE
 * ALONG WITH THIS PROGRAM; IF NOT, WRITE TO THE FREE SOFTWARE
 * FOUNDATION, INC., 59 TEMPLE PLACE - SUITE 330, BOSTON, MA  
 * 02111-1307, USA.

 */

#include "proxylb.h"

#define MAXFD                   64
#define TRANSFER_BUFFER_SIZE    10000
#define ACCEPT                  0
#define DENY                    1
#define DEFAULT_DENY            2
#define NO_RULE                 3

struct proxy_t {
  int fd;
};

extern int   optind;
extern char *optarg;

static int show_help;
static int show_version;
static int nodaemon;
static int debug;

char *conf_file = NULL;

/* SETUP LONG_OPTS ARRAY FOR ARGUMENTS */
static struct option const long_opts[] = {
    {"nodaemon" , no_argument      , &nodaemon    , 1  },
    {"debug"    , no_argument      , &debug       , 1  },
    {"srcport"  , required_argument, NULL         , 's'},
    {"conf_file", required_argument, NULL         , 'c'},
    {"srcip"    , required_argument, NULL         , 'S'},
    {"help"     , no_argument      , &show_help   , 1  },
    {"version"  , no_argument      , &show_version, 1  },
    {NULL       , 0, NULL, 0}
};

/* void CLOSE()
 * 
 * WRAPPER FOR CLOSE() FUNCTION, DOES ERROR CHECKING.
 */
void Close(int descriptor) {
    if (close (descriptor) == -1)
        syslog (LOG_ERR, "Close problems %m");
}

/* VOID PRINTUSAGE()
 *
 * PRINTS OUT USAGE DIRECTIONS FOR PROXY.
 */
void
PrintUsage (int n) {

    if (n == 1) {

        fprintf (stderr, "Try `proxy --help' for more information.\n");

    } else if (n == 2) {

        fprintf (stderr, "\nUsage: proxylb OPTIONS\n"
           "Proxy and load balance connections from source to dest\n\n"
           "-s, --srcport            Source port        (required)\n"
           "-S, --srcip              Source IP          (optional)\n"
           "-c, --config             Configuration file (optional)\n"
           "    if not given ./config.txt will be used\n"
           "-n, --nodaemon           Run proxy _NOT_ as UNIX daemon (defaults to daemon mode)\n"
           "-d, --debug              Turn on syslog debugging\n"
           "    --help               display this help and exit\n"
           "    --version            output version information and exit\n\n"
           "This is " PROXYLB_VERSION "\n\n"

           "Report bugs, success stories to Miroslav Madzarevic, githubcode@osadmin.com\n"
           "This is modified work as of May 1st 2002, "
           "based originally on proxy-2.2.4 program by sparlin@openpro.org "
           "using GPL licence version 2.\n"
           "The licence can be found in LICENCE file in this programs directory.\n\n"
        );
    } else {
        fprintf (stderr, PROXYLB_VERSION "\n" );
    }
    exit (0);
}

/*
 * INT GETLARGESTSOCKET()
 *
 * COMPARES THE TWO SOCKET DESCRIPTORS AND RETURNS THE ONE WITH THE
 * LARGEST VALUE, PLUS ONE
 */
int GetLargestSocket(int sd1, int sd2) {
    if (sd1 > sd2)
        return(sd1 + 1);
    else
        return(sd2 + 1);
}

/*
 * VOID * PROXY()
 *
 * READS DATA FROM SOCKET DESCRIPTOR CREATED BY NEW CLIENT 
 * AND PASSES IT TO SOCKET DESCRIPTOR WE CREATE. SYNCHRONOUS
 * I/O MULTIPLEXING IS DONE VIA SELECT. SEE SELECT(2) FOR DETAILS.
 */
void *
proxy (void *ptr) {

    struct proxy_t *pt;
    int    sd, sdOut;
    char   buf[TRANSFER_BUFFER_SIZE];
    int    nbytes;
    fd_set rset;
    int    selectval;
    int    iMaxSocket;
    char   *ip;
    char   *port;
    int    nr_nodes = 0; 
    struct node *head = NULL;
    // mire

    // set seed for random numbers

    srand((unsigned int) time(NULL));
    
    debug==1?syslog(LOG_NOTICE,"%s\n","proxy() started"):NULL; 

    read_config(&head,&nr_nodes,conf_file,debug);
    
    // end-mire 

    // call get_random_connection_params 
    get_random_connection_params(&ip,&port,head,nr_nodes,debug);
    
    pt = (struct proxy_t *) ptr;
    sd = pt->fd;
    free(ptr);

    /*
     * CONNECT TO REQUESTED HOST
     */

    debug==1?syslog(LOG_INFO,"about to connect to '%s':'%s'",ip,port):NULL;
    sdOut = tcp_connect(ip, port);
    debug==1?syslog(LOG_INFO,"passed tcp connect"):NULL;

    iMaxSocket = GetLargestSocket(sd, sdOut);

    /*
     * SET UP A SELECT LOOP FOR READING AND WRITING
     */
    for (;;) {
        FD_SET (sd   , &rset);
        FD_SET (sdOut, &rset);
        selectval = select (iMaxSocket, &rset, NULL, NULL, NULL);
        if (selectval == -1) {
            syslog (LOG_ERR, "select(): %m");
            Close(sdOut);
            Close(sd);
            break;
    	}
      
        /*
        * IS CLIENT READABLE? IF SO, READ FROM CLIENT, 
        * WRITE TO DESTINATION.
        */
        if (FD_ISSET (sd, &rset)) {
            nbytes = read(sd, &buf, TRANSFER_BUFFER_SIZE);
            if (nbytes <= 0) {
                Close(sdOut);
                Close(sd); 
                break;
	        }
            if (write (sdOut, &buf, nbytes) != nbytes)
                syslog(LOG_ERR, "error on write %m");
	        nbytes -= nbytes;
	    }
      
        /*
         * IS DESTINATION READABLE? IF SO, READ FROM DESTINATION, 
         * WRITE TO CLIENT. 
        */

        if (FD_ISSET (sdOut, &rset)) {
            nbytes = read(sdOut, &buf, TRANSFER_BUFFER_SIZE);
            if (nbytes <= 0) {
                Close(sdOut);
                Close(sd);
                break;
            }
            if (write (sd, &buf, nbytes) != nbytes)
                syslog(LOG_ERR, "error on write %m");
            nbytes -= nbytes;
        }
    } // end of for eternal read loop

    // mire
    kill_list(&head,&nr_nodes,debug);
    free(ip);
    free(port);
    debug==1?syslog(LOG_NOTICE,"proxy() %s","finished"):NULL;

    // end-mire
    return NULL;
}

/*
 * INT DAEMON_INIT()
 *
 * CAUSES PROXY TO RUN AS A UNIX DAEMON
 */
int
daemon_init (const char *pname) {
    int   i;
    pid_t pid;

    if ((pid = fork ()) != 0)
        exit (0);

    setsid ();

    signal (SIGHUP, SIG_IGN);

    if ((pid = fork ()) != 0)
        exit (0);

    // no no no chdir ("/");

    umask (0);

    /* 
     * CLOSE THE FIRST 64 DESCRIPTORS INHERITED FROM THE PROCESS
     * THAT EXECUTED THE DAEMON. THERE ARE PROBABLY NOT 64 OPEN
     * DESCRIPTORS, SO I DON'T CHECK THE RETURN VALUE OF CLOSE().
     * SEE "UNIX NETWORK PROGRAMMING VOLUME 1" FOR MORE INFORMATION
     * ON THIS SUBJECT... (PAGE 337 PARAGRAPH 3)
     */
    for (i = 0; i < MAXFD; i++)
        close(i);

    return 0;
}

/*
 * INT FILTERIP()
 *
 * BASIC IP FILTERING
 */
int
FilterIP (char *clientip) {
    FILE *access;
    char str[MAXLINE];
    int  deny = 0 , accept = 0;
    char *deny_str, *access_str ;
    int  i;

    access = fopen ("/etc/proxy.filters", "r");

    if (access == NULL) {
        debug==1?syslog (LOG_ERR, "fopen problems on filters file /etc/proxy.filters (harmless) %m"):NULL;
        return 0;
    }

    access_str = (char *) malloc (sizeof (char) * 25);
    sprintf (access_str, "ACCEPT.%s", clientip);

    deny_str = (char *) malloc (sizeof (char) * 25);
    sprintf (deny_str, "DENY.%s", clientip);

    while (fgets (str, MAXLINE, access) != NULL) {
        if (str[0] == '#') 
            continue;

        for (i = 0; str[i] != '\n'; i++);
        str[strlen (str) - 1] = '\0';

        for (i = 0; str[i] != '.'; i++)
            str[i] = toupper (str[i]);

        if (strstr (access_str, str) != NULL) {
            free (access_str);
            free (deny_str);
            return ACCEPT;
        }

        if (strstr (deny_str, str) != NULL) {
            free (access_str);
            free (deny_str);
            return DENY;
        }

        if (strcmp (str, "ACCEPT;") == 0)
	        accept++;
        else if (strcmp (str, "DENY;") == 0)
        	deny++;
    }

    free   (access_str);
    free   (deny_str);
    fclose (access);

    if (deny > 0)
        return DEFAULT_DENY;
    else if (accept > 0)
        return ACCEPT;
    else
        return NO_RULE;
}

/*
 * STATIC VOID FINISH()
 *
 * FUNCTION TO HANDLE SIGNAL GENERATED BY CTRL-C
 */
void
finish (int sig) {
    pthread_exit(NULL);
}

/*
 * INT MAIN()
 */
int
main (int argc, char *argv[]) {
    int       listenfd, connfd;
    socklen_t clilen, clientaddrlen;
    socklen_t addrlen;
    struct    sockaddr *cliaddr;
    struct    sockaddr *clientaddr;
    struct    proxy_t *ptr;
    pthread_t ptConnection;
    int   FilterCheck;
    int   c;
    char  *srcip   = NULL;
    char  *srcport = NULL;

    (void) signal (SIGINT, finish);

    while ((c = getopt_long (argc, argv, "nds:S:c:", long_opts, NULL)) != -1) {
        switch (c) {
        
            case 0:
        	  break;

        	case 'n':
        	  nodaemon = 1;
        	  break;

        	case 'd':
        	  debug = 1;
        	  break;

        	case 's':
        	  srcport = optarg;
        	  break;

        	case 'c':
        	  conf_file = optarg;
        	  break;

        	case 'S':
        	  srcip = optarg;
        	  break;

        	default:
        	  PrintUsage (1);
              exit(1);
    	}
    }

    if (show_help)
        PrintUsage (2);

    if (show_version)
        PrintUsage (3);

    if (argc < 3) {
        fprintf (stderr, "proxylb: too few arguments\n");
        PrintUsage (1);
    }

    if (!srcport)
        PrintUsage(1);

    /* PRINT STARTUP MESSAGE */
    fprintf (stderr, "\nproxylb started: listening on port [%s]\n\n", srcport);

    /* RUN PROXY AS UNIX DAEMON */
    if (!nodaemon)
        daemon_init (argv[0]);

    if (srcip)
        listenfd = tcp_listen (srcip, srcport, &addrlen);
    else
        listenfd = tcp_listen (NULL, srcport, &addrlen);

    if ( (cliaddr = malloc(addrlen)) == NULL)
        syslog(LOG_ERR, "memory allocation error %m");

    // eternal socket accept loop
    for (;;) {

        ptr = (struct proxy_t *) malloc (sizeof (struct proxy_t));

        clilen = addrlen;
        connfd = accept (listenfd, cliaddr, &clilen);

        if (connfd == -1)
            continue;

        ptr->fd = connfd;

        if ((clientaddr = malloc (MAXSOCKADDR)) == NULL)
            syslog (LOG_ERR, "memory allocation error %m");

        clientaddrlen = MAXSOCKADDR;

        if (getpeername (ptr->fd, clientaddr, &clientaddrlen) < 0)
    	    syslog (LOG_ERR, "getpeername error %m");

        /*
         * BASIC IP FILTERING
         */
        FilterCheck = FilterIP ((char *) Sock_ntop (clientaddr, clientaddrlen));

        if (FilterCheck > 0) {

            switch (FilterCheck) {
                case DENY:
            		debug==1?syslog (LOG_INFO, "found deny rule for %s, dropping connection!",
        	    		Sock_ntop (clientaddr, clientaddrlen)):NULL;
                    break;
                case DEFAULT_DENY:
                	debug==1?syslog (LOG_INFO, "%s matched default policy of DENY, " 
                        "dropping connection!",
        	    		Sock_ntop (clientaddr, clientaddrlen)):NULL;
                     break;
                case NO_RULE:
               		debug==1?syslog (LOG_ERR, "no rule matching %s and no default policy, "
                		"dropping!", Sock_ntop (clientaddr, clientaddrlen)):NULL;
                    break;
            }

            Close (ptr->fd);
        } else {
            debug==1?syslog (LOG_INFO, "allowing connection from %s",
		        Sock_ntop (clientaddr, clientaddrlen)):NULL;
        }

        free(clientaddr);

        if (pthread_create (&ptConnection, NULL, proxy, (void *) ptr) != 0) {
    	    syslog (LOG_ERR, "pthread creation problems %m");
	        return 1;
        }

        if ( (pthread_detach(ptConnection)) != 0)
    	    syslog(LOG_ERR, "pthread_detach %m");
    }
} // end of main

// _END_

